### PR TITLE
fix(core): access popover on image using keyboard

### PR DESCRIPTION
**Pull request by sKudum**
#### Please provide a link to the associated issue.
Closes SAP/fundamental-ngx #3139 

#### Please provide a brief summary of this pull request.

The ‘image’ buttons highlighted in popover will be reachable by keyboard as well.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [n/a] update `README.md`
- [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: libs/core/src/lib/popover/popover.component.ts [hunk 1]
```diff
@@ -146,7 +146,8 @@ export class PopoverComponent extends BasePopoverClass implements AfterViewInit,
 
     /** Handler for alt + arrow down keydown */
     triggerKeyDownHandler(event: KeyboardEvent): void {
-        if (KeyUtil.isKeyCode(event, DOWN_ARROW) && event.altKey && !this.disabled) {
+        if ((KeyUtil.isKeyCode(event, DOWN_ARROW) && event.altKey && !this.disabled) ||
+         (KeyUtil.isKeyCode(event, ENTER) || KeyUtil.isKeyCode(event, SPACE))) {
```
Comment (Betrozov) [inline libs/core/src/lib/popover/popover.component.ts:150]:
```suggestion
         KeyUtil.isKeyCode(event, [ENTER, SPACE])) {
```

File: apps/docs/src/app/core/component-docs/popover/examples/popover-simple/popover-example.component.scss [hunk 2]
```diff
@@ -1,4 +1,26 @@
-.fd-docs-flex-display-helper {
+$fd-focus-outline-offset: 0.0625rem;
+
+@mixin fd-fiori-focus($offset: -0.1875rem) {
+    &:focus,
+    &.is-focus {
+      outline-offset: $offset;
+      outline-width: var(--sapContent_FocusWidth);
+      outline-color: var(--sapContent_FocusColor);
+      outline-style: var(--sapContent_FocusStyle);
```
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-simple/popover-example.component.scss:9]:
```suggestion
    outline-width: .0625rem;
    outline-width: var(--sapContent_FocusWidth, .0625rem);
    outline-color: #000;
    outline-color: var(--sapContent_FocusColor, #000);
    outline-style: dotted;
    outline-style: var(--sapContent_FocusStyle, dotted);
```

File: apps/docs/src/app/core/component-docs/popover/examples/popover-programmatic/popover-programmatic-open-example.component.scss [hunk 3]
```diff
@@ -6,3 +6,19 @@
...
 }
+$fd-focus-outline-offset: 0.0625rem;
+
+@mixin fd-fiori-focus($offset: -0.1875rem) {
+    &:focus,
+    &.is-focus {
+      outline-offset: $offset;
+      outline-width: var(--sapContent_FocusWidth);
+      outline-color: var(--sapContent_FocusColor);
+      outline-style: var(--sapContent_FocusStyle);
```
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/popover/examples/popover-programmatic/popover-programmatic-open-example.component.scss:17]:
```suggestion
    outline-width: .0625rem;
    outline-width: var(--sapContent_FocusWidth, .0625rem);
    outline-color: #000;
    outline-color: var(--sapContent_FocusColor, #000);
    outline-style: dotted;
    outline-style: var(--sapContent_FocusStyle, dotted);
```

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Review by KevinOkamoto — APPROVED** [review]

**Review by Betrozov — APPROVED** [review]

**Review by Seamoo13 — APPROVED** [review]

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]
- `outline-offset` is not supported on IE11, we have another mixin for applying a fake focus (used in buttons and list)
- We agreed to not have css in Fundamental-ngx, I am not allowed to approve PRs with css code in this library.

**Comment by droshev** [comment]
Closing old PRs - feel free to reopen it if/when needed

**Comment by sKudum** [comment]
Reopening it to address reviewers comments.

**Comment by sKudum** [comment]
> * `outline-offset` is not supported on IE11, we have another mixin for applying a fake focus (used in buttons and list).
          
   
Hi @InnaAtanasova , 
    1).As mentioned outline-offset is removed from styles.

   2).Tried using fake focus referring to list and button but look as feel is not as expected attached screen shots of out put by using fake focus.




Code sample:






> * We agreed to not have css in Fundamental-ngx, I am not allowed to approve PRs with css code in this library.

    These styles are added in documentation examples as a11y team is looking for these feature even though the respective components such as avatar and icon doesn't provide the focus (as per Fiori3 documents).

**Comment by InnaAtanasova** [comment]
> > * `outline-offset` is not supported on IE11, we have another mixin for applying a fake focus (used in buttons and list).
> 
> Hi @InnaAtanasova ,
> 
> ```
> 1).As mentioned outline-offset is removed from styles.
> 
> 2).Tried using fake focus referring to list and button but look as feel is not as expected attached screen shots of out put by using fake focus.
> ```
> 
> 
> 
> Code sample:
> 
> 
> > * We agreed to not have css in Fundamental-ngx, I am not allowed to approve PRs with css code in this library.
> 
> ` These styles are added in documentation examples as a11y team is looking for these feature even though the respective components such as avatar and icon doesn't provide the focus (as per Fiori3 documents).`
> 
> Please do let me know your suggestions on this will update accordingly.Thank You.

In my opinion it makes no sense to make the components in the documentation accessible if they are not accessible by default. When the user uses these components in their applications they won't get the styling from our documentation example, which means that they won't be compliant to the a11y rules. Avatar and icons by default do not have focus because focus indicates some sort of an action associated with them.  The moment they are used as "controls" , in this case to open the popover body they should get a focus. Here are some examples from UI5:
Screen Shot 2021-06-09 at 1 12 36 PM
Screen Shot 2021-06-09 at 1 13 28 PM
Screen Shot 2021-06-09 at 1 12 40 PM

Check the last icon from here: https://sapui5.hana.ondemand.com/#/entity/sap.ui.core.Icon/sample/sap.ui.core.sample.Icon
and the first row of Avatars here: https://sapui5.hana.ondemand.com/#/entity/sap.f.Avatar/sample/sap.f.sample.Avatar

As for the focus line and the problem with focus line covering the avatar, this specific fake focus mixin was created for list and/or button, you can always create a new one for Avatar and play with the left, top, bottom, right position values.

**Comment by InnaAtanasova** [comment]
In my opinion when Avatar is placed inside a popover control it should not have the `role="img"` but `role="button"`, because as per MDN:
 _"The button role should be used for clickable elements that trigger a response when activated by the user. Adding role="button" will make an element appear as a button control to a screen reader. This role can be used in combination with the aria-pressed attribute to create toggle buttons."_ 

More info here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role

Same for icons, images, basically all the items that can become a trigger for action. You can use css specificity and add the focus outline and the pointer cursor when the class comes together with a role: 
```
.fd-avatar[role="button"] {...}
```

Hope this helps. Let me know if more info is required.

**Comment by sKudum** [comment]
Hi @InnaAtanasova , will go with approach 2. 

Please suggest what should I do for now.
1). Until styles are published can I have styles in lib later will update styles.
2).Close this PR and reopen it once styles are published and do the required changes.

**Comment by InnaAtanasova** [comment]
> Hi @InnaAtanasova , will go with approach 2.
> 
> Please suggest what should I do for now.
> 1). Until styles are published can I have styles in lib later will update styles.
> 2).Close this PR and reopen it once styles are published and do the required changes.

Hey Sushma,
you need to close this PR and open a PR in fund-styles. Once it's merged you can use the RC and continue in ngx.

**Comment by sKudum** [comment]
Closing it as discussed mentioned above.

**Pull request closed without merging**